### ISSUE TITLE
ValueChangeListener prevents client-side click listener

### ISSUE THREAD

**Issue by probert94** [vaadin/flow-components#5006]
### Description

While trying to create a minimal reproducible example for another issue, I stumbled upon a really strange issue.
Basically I have a `Dialog` with a `TextField`. I register a `ValueChangeListener` for the `TextField`, that opens another `Dialog`.
Now I noticed, that if the `ValueChangeEvent` is triggered by a click outside of the `TextField`, the `ClickEvent` itself does not get fired, not even on the client.
If the `ValueChangeListener` takes a little longer (I used `Thread.sleep(100)` to simulate that), the `ClickEvent` gets fired normaly.

The issue does not occur in Firefox, but in Chrome, Edge and Opera. So it seems to be Chromium-Related.

### Expected outcome

The `ClickEvent` should fire on the client and be sent to the server if needed

### Minimal reproducible example

```java
@PageTitle("Main")
@Route(value = "")
public class MainView extends HorizontalLayout {
  public MainView() {
    setPadding(true);
    Button btnOpenDlg = new Button("Open", e -> openDlg1());
    add(btnOpenDlg);
  }
  private static void openDlg1() {
    Dialog dlg = new Dialog();
    dlg.setModal(true);
    dlg.setCloseOnOutsideClick(true);
    HorizontalLayout toolBar = new HorizontalLayout();
    Span title = new Span("Title");
    toolBar.addAndExpand(title);
    dlg.add(toolBar);
    VerticalLayout content = new VerticalLayout();
    TextField tfValue = new TextField("Value");
    tfValue.addValueChangeListener(e -> openDlg2(e.getValue()));
    content.add(tfValue);
    dlg.add(toolBar);
    dlg.add(content);
    dlg.addOpenedChangeListener(e -> System.out.println(e.isOpened() ? "Opened" : "Closed"));
    dlg.open();
  }
  private static void openDlg2(String value) {
    Dialog dlg = new Dialog();
    dlg.setModal(true);
    Div content = new Div();
    content.setText(value);
    dlg.add(content);
    dlg.open();
  }
}
```

### Steps to reproduce

1. Dowload a Vaadin starter (I used version 24.0.5)
2. Copy the above code into `MainView.java`
3. Start the project and open a `Dialog`
4. To verify if a `ClickEvent` is fired open the developer tools and insert the following command into the console:
   `document.body.addEventListener("click", e => console.warn(e))`
5. Insert some text into the `TextField` and click somewhere outside of it.

Now a new `Dialog` should open but there shouldn't be a new `ClickEvent` in the console

### Environment

Vaadin: 24.0.5
Flow: 24.0.6
Java: Eclipse Adoptium 17.0.6
OS: amd64 Windows 11 10.0

### Browsers

Chrome, Edge

**Comment by probert94** [comment]
> The fact that document.body.addEventListener("click", e => console.warn(e)) doesn't log anything to the console isn't really anything that is problematic on its own

That shows, that the opening of the `Dialog` (which happens over the server) is able to prevent a client side click event.
Of course the `console.warn` is nothing important, it is just here to show a potential problem.

> It would be helpful if you could explain what you are trying to achieve, and what part is problematic for your application

I am currently not facing a problem related to this, I just discovered this potential problem while trying to create a minimal reproducible example for another issue.  
I created this issue because the behavior might be problematic in some cases and because I am curious if anyone has an explaination for this.

> The only issue I see is that the second dialog does not stay open consistently

Basically the click event get's lost, cause of the opening of a `Dialog`. So whatever logic the click-event should trigger, it won't be executed. Maybe there are even other events affected by this.

As a sidenote:
I noticed this behavior while trying to reproduce the referenced issue ( #4997 ). I noticed that the "value-change" event and the "click" event are sent to the server in two separate requests, even if they are triggered by the same user interaction.
If they were sent to the server in the same request, this problem as well as the one in the referenced issue would probably be solved. Not sure if that is possible though.

**Comment by sissbruecker** [comment]
> I noticed that the "value-change" event and the "click" event are sent to the server in two separate requests, even if they are triggered by the same user interaction.
If they were sent to the server in the same request, this problem as well as the one in the referenced issue would probably be solved. Not sure if that is possible though.

That is explainable by the order of events that occur when clicking outside of the text field:
- mousedown - Mouse button is pressed down
- blur - mousedown causes the input field to blur
- mouseup - Mouse button is released
- click - Finally triggered after both mousedown and mouseup

Releasing the mouse button is something that can take some time, at that point the blur event has already been sent to the server. I honestly don't know how the Flow client batches events, but it doesn't (and probably shouldn't) wait that long.

---

Apart from that I really don't see what we should fix here. I can't see a use-case where you need to listen to a click event on the document that is part of the same user action that causes a dialog to open. I'll close the issue, feel free to create a new one if you have a use-case that is impacted by this behavior.

**Comment by probert94** [comment]
> That is explainable by the order of events that occur

Thank you for clearing that up, that makes sense now.
That also explains why the click event isn't fired on Chrom. Chrome probably doesn't fire click events that are not on the same element es the related `mousedown`. Since the `Dialog` opens before the `click` event occurs, it probably "hides" other elements behind it's backdrop.

>  I can't see a use-case where you need to listen to a click event on the document that is part of the same user action that causes a dialog to open.

The `Dialog` (and other `Overlay` components) use a `outside-click` listener to close, if `closeOnClickOutside` is set which won't be triggered in this cases.
In my case the server takes long enough to process potential `value-change` events so this is not an issue for me.

Thank you
